Recently added WITH_ASAN check in memory.c is misplaced

Investigating #4980 I found that I could not trigger ASAN's out-of-bounds checks on buffer got from `mem_alloc_tiny()` easily while I had unchecked `strcpy()`. A huge overflow is required to trigger ASAN.

I don't know if `mem_alloc_tiny()` is used with `--fuzz`. If so, it should degrade quality of fuzzing greatly.

Example from `src/odf_common_plug.c`:
```c
		static char *buf = NULL;
[...]
		if (!buf)
			buf = mem_alloc_tiny(3*1024, 4);
		cp1 = buf;
[...]
		strcpy(cp1, cp3);
```

~30kb hash does not trigger it:
```
$ python -c 'print "$sxc$*1*2*3*4*5*6*7*8*9*10*11*" + "a" * 30000' > t.pw && ./run/john t.pw --show=formats > /dev/null
```

~40kb hash triggers ASAN:
```
$ python -c 'print "$sxc$*1*2*3*4*5*6*7*8*9*10*11*" + "a" * 40000' > t.pw && ./run/john t.pw --show=formats > /dev/null
=================================================================
==15673==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x63100261c800 at pc 0x7f0022dfe274 bp 0x7fffc2c833f0 sp 0x7fffc2c82ba0
WRITE of size 40001 at 0x63100261c800 thread T0
    #0 0x7f0022dfe273 in __interceptor_strcpy ../../../../src/libsanitizer/asan/asan_interceptors.cc:391
    #1 0x55679b0e1634 in odf_prepare /home/a/d/jp/22_fuzzbug/john/src/odf_common_plug.c:235
```

I just rebuilt from repo and rechecked: same behaviour is in place.

```
$ ./configure --enable-asan --disable-openmp --enable-fuzz
```

I modified `memory.c`:
```c
#if defined(DEBUG)
#warning mywarn: debug
#else
#warning mywarn: no debug
#endif

#if defined(WITH_ASAN) && !defined(DEBUG)
#warning mywarn: with_asan
#define DEBUG
#else
#warning mywarn: no with_asan
#endif

#if defined(DEBUG)
#warning mywarn: debug 2
#else
#warning mywarn: no debug 2
#endif
```

```
$ make -s clean && make -sj4
memory.c:24:2: warning: #warning mywarn: no debug [-Wcpp]
 #warning mywarn: no debug
  ^~~~~~~
memory.c:31:2: warning: #warning mywarn: no with_asan [-Wcpp]
 #warning mywarn: no with_asan
  ^~~~~~~
memory.c:37:2: warning: #warning mywarn: no debug 2 [-Wcpp]
 #warning mywarn: no debug 2
  ^~~~~~~
ar: creating aes.a
ar: creating poly1305-donna.a
ar: creating ed25519-donna.a
ar: creating secp256k1.a
```
Plus same set of 3 warnings repeats 3 times.

So it looks like `WITH_ASAN` is not set.

--- a/src/krb5_db_fmt_plug.c
+++ b/src/krb5_db_fmt_plug.c
@@ -253,7 +253,8 @@ static void *get_binary(char *ciphertext)
 		p = ciphertext + TAG_LENGTH_3;
 	}
 
-	if (!out) out = mem_alloc_tiny(binary_size, MEM_ALIGN_WORD);
+	/* 32 is max possible binary_size above */
+	if (!out) out = mem_alloc_tiny(32, MEM_ALIGN_WORD);
 	p = strstr(p, "$") + 1;
 
 	for (; i < binary_size; i++) {

--- a/src/listconf.c
+++ b/src/listconf.c
@@ -86,7 +86,7 @@
 #include "mask_ext.h"
 #include "john.h"
 #include "version.h"
-#include "listconf.h" /* must be included after version.h */
+#include "listconf.h" /* must be included after version.h and misc.h */
 
 #ifdef NO_JOHN_BLD
 #define JOHN_BLD "unk-build-type"

--- a/src/memory.c
+++ b/src/memory.c
@@ -18,15 +18,15 @@
 #include <malloc.h>
 #endif
 
-#if defined(WITH_ASAN) && !defined(DEBUG)
-#define DEBUG
-#endif
-
 #include "arch.h"
 #include "misc.h"
 #include "memory.h"
 #include "common.h"
 #include "johnswap.h"
+
+#if defined(WITH_ASAN) && !defined(DEBUG)
+#define DEBUG
+#endif
 
 #if (defined (_MSC_VER) || HAVE___MINGW_ALIGNED_MALLOC)
 char *strdup_MSVC(const char *str)

--- a/src/options.c
+++ b/src/options.c
@@ -50,7 +50,7 @@
 #include "prince.h"
 #endif
 #include "version.h"
-#include "listconf.h" /* must be included after version.h */
+#include "listconf.h" /* must be included after version.h and misc.h */
 #include "jumbo.h"
 
 struct options_main options;

